File: solutions_2453486_1/C++/jplotzke/tomek.cpp
#include <cstdio>
#include <iostream>
#include <string>
using namespace std;

string check_line(char board[4][4], char player, int y, int x, int dy, int dx){
	if (player == 'T'){
		if (check_line(board, 'X', y, x, dy, dx) == "X")
			return "X";
		return check_line(board, 'O', y, x, dy, dx);
	}
	int count; string ans = "";
	ans += player;
	for (count = 0; y >= 0 && y < 4 && x >= 0 && x < 4 && (board[y][x] == player || board[y][x] == 'T'); y += dy, x += dx, ++count);
	return (count==4 ? ans : ".");
}

string check_board(char board[4][4]){
	string ans;
	for (int y = 0; y < 4; ++y)
		if ((ans = check_line(board, board[y][0], y, 0, 0, 1)) != ".")
			return ans + " won";
	for (int x = 0; x < 4; ++x)
		if ((ans = check_line(board, board[0][x], 0, x, 1, 0)) != ".")
			return ans + " won";
	if ((ans = check_line(board, board[0][0], 0, 0, 1, 1)) != ".")
		return ans + " won";
	if ((ans = check_line(board, board[0][3], 0, 3, 1, -1)) != ".")
		return ans + " won";
	for (int y = 0; y < 4; ++y)
		for (int x = 0; x < 4; ++x)
			if (board[y][x] == '.')
				return "Game has not completed";
	return "Draw";
}

int main(){
	freopen("tomek.in", "r", stdin);
	freopen("tomek.out", "w", stdout);
	int run; char board[4][4]; string line;
	cin >> run;
	for (int a = 1; a <= run; ++a){
		getline(cin, line);
		for (int y = 0; y < 4; ++y){
			getline(cin, line);
			for (int x = 0; x < 4; ++x) board[y][x] = line[x];
		}
		cout << "Case #" << a << ": " << check_board(board) << endl;
	}
	return 0;
}

File: solutions_2453486_1/java/SiqiLin/Tomek.java
import java.io.*;

public class Tomek {
    private static void checkBoard(String[] board, BufferedWriter bw, int caseNumber) throws IOException {
        int emptyCount = 0;
        
        for (int x = 0; x < 4; x++) {
            int Xcount = 0;
            int Ocount = 0;

            for (int y = 0; y < 4; y++) {
                char c = board[x].charAt(y);
                if (c == '.') {
                    emptyCount++;
                } else if (c == 'T') {
                    Xcount++;
                    Ocount++; 
                } else if (c == 'X') {
                    Xcount++;
                } else if (c == 'O') {
                    Ocount++;
                } 
            }
            if (Xcount == 4) {
                bw.write("Case #" + caseNumber + ": X won");
                bw.newLine();
                return;
            } else if (Ocount == 4) {
                bw.write("Case #" + caseNumber + ": O won");
                bw.newLine();
                return;
            }
        } 

        for (int y = 0; y < 4; y++) {
            int Xcount = 0;
            int Ocount = 0;

            for (int x = 0; x < 4; x++) {
                char c = board[x].charAt(y);
                if (c == 'T') {
                    Xcount++;
                    Ocount++; 
                } else if (c == 'X') {
                    Xcount++;
                } else if (c == 'O') {
                    Ocount++;
                }
            }
            if (Xcount == 4) {
                bw.write("Case #" + caseNumber + ": X won");
                bw.newLine();
                return;
            } else if (Ocount == 4) {
                bw.write("Case #" + caseNumber + ": O won");
                bw.newLine();
                return;
            }
        }

        int Xcount = 0;
        int Ocount = 0;

        for (int i = 0; i < 4; i++) {
            char c = board[i].charAt(i);
            if (c == 'T') {
                Xcount++;
                Ocount++; 
            } else if (c == 'X') {
                Xcount++;
            } else if (c == 'O') {
                Ocount++;
            }
        }

        if (Xcount == 4) {
            bw.write("Case #" + caseNumber + ": X won");
            bw.newLine();
            return;
        } else if (Ocount == 4) {
            bw.write("Case #" + caseNumber + ": O won");
            bw.newLine();
            return;
        }

        Xcount = 0;
        Ocount = 0;

        for (int i = 0; i < 4; i++) {
            char c = board[3 - i].charAt(i);
            if (c == 'T') {
                Xcount++;
                Ocount++; 
            } else if (c == 'X') {
                Xcount++;
            } else if (c == 'O') {
                Ocount++;
            }
        }

        if (Xcount == 4) {
            bw.write("Case #" + caseNumber + ": X won");
            bw.newLine();
            return;
        } else if (Ocount == 4) {
            bw.write("Case #" + caseNumber + ": O won");
            bw.newLine();
            return;
        }

        if (emptyCount == 0) {
            bw.write("Case #" + caseNumber + ": Draw");
        } else {
            bw.write("Case #" + caseNumber + ": Game has not completed");
        }
        bw.newLine();
    }

    public static void main(String[] args) {
        BufferedReader br = null;
        BufferedWriter bw = null;

        try {
            File outputFile = new File("./Tomek_out.txt");
            br = new BufferedReader(new FileReader("./Tomek_in.txt"));

            if (!outputFile.exists()) {
              outputFile.createNewFile();
            }

            bw = new BufferedWriter(new FileWriter(outputFile));

            String currentLine;

            currentLine = br.readLine();
            int T = Integer.valueOf(currentLine);

            for (int t = 1; t <= T; t++) {
                String[] board = new String[4];
                for (int j = 0; j < 4; j++) {
                    currentLine = br.readLine();
                    board[j] = currentLine;
                }
                checkBoard(board, bw, t);
                br.readLine(); // Get rid of extra line;
            }
            bw.flush();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                if (br != null) {
                    br.close();
                }
                if (bw != null) {
                    bw.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            } 
        }
    }
}
